src/go/pkg/controller/approllout/controller.go

```diff
@@ -411,23 +446,49 @@ func (r errRobotSelectorOverlap) Error() string {
 // generateChartAssignments returns a list of all cloud and robot ChartAssignments
 // for the given app, its rollout, and set of robots.
 func generateChartAssignments(
-	app *apps.App,
+	al []apps.App,
+	robots []registry.Robot,
 	rollout *apps.AppRollout,
-	allRobots []registry.Robot,
 	baseValues chartutil.Values,
 ) ([]*apps.ChartAssignment, error) {
+
 	var (
-		cas   []*apps.ChartAssignment
-		comps = app.Spec.Components
+		// Different entries might request different app versions. This map
+		// is used to only retrieve them once.
+		appVersions = map[string]*apps.App{}
+		cas         []*apps.ChartAssignment
 		// Robots that matched selectors for the rollout and which will be
 		// passed to the cloud chart.
 		selectedRobots = map[string]*registry.Robot{}
 	)
+
+	for _, app := range al {
+		v, ok := app.Labels[labelAppVersion]
+		if !ok {
+			// If only app-name is defined, it is an unversioned app.
+			v = ""
+		}
+		if _, ok := appVersions[v]; ok {
+			log.Printf("App %q version %q already known. Going to ignore App Object %q for the same app/version", rollout.Spec.AppName, v, app.Name)
+		} else {
+			// only add to map if this is the first time we're adding this
+			// version for the app. The validator should ensure that this
+			// overwrite cannot happen, but apps might be legacy, or have
+			// bypassed validation.
+			appVersions[v] = &app
+		}
+	}
 	for _, rcomp := range rollout.Spec.Robots {
-		robots, err := matchingRobots(allRobots, rcomp.Selector)
+		robots, err := matchingRobots(robots, rcomp.Selector)
 		if err != nil {
 			return nil, errors.Wrap(err, "select robots")
 		}
+		// map is populated by for all the rcomp.Version, no need to check ok
+		app, ok := appVersions[rcomp.Version]
+		if !ok {
+			return nil, fmt.Errorf("no App %q (Version: %q) found", rollout.Spec.AppName, rcomp.Version)
+		}
+		comps := app.Spec.Components
```

Comment:
I don't think this survices the scope of the loop. See eg. https://stackoverflow.com/questions/33823209/variable-in-for-loop-undefined-outside